ci/src/Registry/Index.purs

```
@@ -0,0 +1,128 @@
+  according to the following format:
+
+    * Packages with 1 character names are placed in a directory named 1.
+    * Packages with 2 character names are placed in a directory named 2.
+    * Packages with 3 character names are placed in the directory 3/{first-character} where {first-character} is the first character of the package name.
+    * All other packages are stored in directories named {first-two}/{second-two} where the top directory is the first two characters of the package name, and the next subdirectory is the third and fourth characters of the package name. For example, prelude would be stored in a file named pr/el/prelude.
+-}
+getIndexDir :: PackageName -> FilePath
+getIndexDir (PackageName packageName) = case String.length packageName of
+  0 -> unsafeCrashWith "Invalid PackageName"
+  1 -> "1/"
+  2 -> "2/"
+  3 -> "3/" <> String.take 1 packageName <> "/"
+  _ -> String.take 2 packageName <> "/" <> String.take 2 (String.drop 2 packageName) <> "/"
+
+getIndexPath :: PackageName -> FilePath
+getIndexPath (PackageName packageName) = getIndexDir (PackageName packageName) <> packageName
+
+-- Collect all manifests for given PackageName
+-- This function must be run from the root of the registry index.
+-- This will return Nothing if:
+--   the file doesn't exist, the file is empty, or if we can't decode the Manifests
+readPackage :: PackageName -> Aff (Maybe (NonEmptyArray Manifest))
+readPackage packageName = do
+  let
+    path = getIndexPath packageName
+
+  contentsResult <- try do
+    contents <- readTextFile ASCII path
+    pure $ hush do
+      jsonLines <- traverse parseJson $ String.split (Pattern "\n") contents
+      traverse decodeJson jsonLines
+
+  pure case contentsResult of
+    Left _ -> Nothing
+    Right Nothing -> Nothing
+    Right (Just arr) -> NEA.fromArray arr
+
+-- This function must be run from the root of the registry index.
+-- TODO: Should we bail on version collisions?
```

Thoughts on this, @f-f? This PR will always write the manifest, even if the manifest already exists in the index file at the given version. We could instead bail out and not write it, or we could throw an exception...I'm not sure if there's some danger to overwriting in the case of version collision.

---

Version collisions should be prevented by the fact that we store them in a `Map`, but if we want to rule them out we can check if the Manifest we're about to write is the same as the one that's there.
Since the index is supposed to be just a cache (that you can throw away) I think it's fine to not worry aboout this too much.